kernel/EigenSolvers/JacobiDavidsonMaxwell.h

```diff
@@ -0,0 +1,144 @@
+/*
+This file forms part of hpGEM. This package has been developed over a number of
+years by various people at the University of Twente and a full list of
+contributors can be found at http://hpgem.org/about-the-code/team
+
+This code is distributed using BSD 3-Clause License. A copy of which can found
+below.
+
+
+Copyright (c) 2018, Univesity of Twenete
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice, this
+list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+this list of conditions and the following disclaimer in the documentation and/or
+other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+may be used to endorse or promote products derived from this software without
+specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef JACOBIDAVIDSONMAXWELLSOLVER_H
+#define JACOBIDAVIDSONMAXWELLSOLVER_H
+
+#include <petsc.h>
+#include <petscksp.h>
+
+#include <slepcbv.h>
+#include <slepceps.h>
+
+#include <iostream>
+#include <math.h>
+#include <cstdlib>
+#include <vector>
+
+#include <chrono>  // For timing
+
+#include <utility>
+#include <valarray>
+
+/**
+ * Solver of the Maxwell equation based on the paper
+ *   Multilevel preconditioned iterative eigensolvers
+ *   for Maxwell eigenvalue problems
+ *   P Arbenz R Geus
+ *   Applied Numerical Mathematics 54 (2005) 107-121
+ *
+ *
+ * Implementation note:
+ *
+ */
+
+namespace hpgem {
+
+namespace EigenSolvers {
+
+class JacobiDavidsonMaxwellSolver final {
```

Comment:
A few notes on the documentation here:

It would be nice to give a short description of the eigenvalue problem that is solved. Something like Ax = lambda x, constraint to C x = 0

There are several variables that are rather cryptically named for someone not having the paper next to the code. It might be nice to describe the content of these:

 - Y = C^H, H = C^Y (needed both for removing the space spanned by C)
 -  V = Eigenvalue search space 
 -  Q/Qt = Converged eigenvalues (I have no clue so far what the difference is between these)
 -  search/res_vect intermediate vectors for the search direction and residual.
 - eta = shift in the correction equation
 - tau = eigenvalue target?